### PR TITLE
feat: Add notifications in the app

### LINKED ISSUE #2289: Add notifications in the application
**Actual Behaviour**

The API supports [notifications](https://open-event-api.herokuapp.com/#notifications) but it is not implemented in the app.

**Expected Behaviour**

It should be implemented.

**Would you like to work on the issue?**

Yes

### PR DESCRIPTION
**Pull request by dr0pdb**
Fixes #2289 

Changes: Added Notifications in the application.

Screenshots for the change: 
videotogif_2018 02 07_00 21 23

The notification is user specific, i.e. it needs the user to be logged in. Hence if the user isn't logged in and tries to open the notifications fragment, he/she is redirected to the login activity.

@iamareebjamal Please review

### CODE REVIEW

File: android/app/src/main/java/org/fossasia/openevent/data/Notification.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+    @PrimaryKey
+    @Id(IntegerIdHandler.class)
+    private int id;
+    private String title;
+    private String message;
+    private boolean isRead;
+    private String receivedAt;
+
+    public Notification() {}
```
[the remark this answers is not in the document]
Comment (srv-twry) [reply android/app/src/main/java/org/fossasia/openevent/data/Notification.java]: It is required. I get a lot of errors of these types: 
```
Error:(92, 30) error: constructor Notification in class Notification cannot be applied to given types;
required: int,String,String,boolean,String
found: no arguments
reason: actual and formal argument lists differ in length
```

> found: no arguments

This requires a no argument constructor.

**Comment by iamareebjamal** [comment]
Instead of globally downloading notifications and using event bus, make it lazy. They should be loaded when a person goes to the notification fragment

**Comment by iamareebjamal** [comment]
We will soon be refactoring our current architecture to be similar for other components

**Comment by dr0pdb** [comment]
@iamareebjamal I am only downloading them when a person goes to the notification fragment. It is still using EventBus though.

**Comment by iamareebjamal** [comment]
Yeah, don't use that

**Comment by dr0pdb** [comment]
Changes: Made the fetching of notifications lazy and removed the use of event bus.

@iamareebjamal Can you help with fixing the build? It says 
```
FAILURE: Build failed with an exception.
* What went wrong:
A problem occurred configuring project ':app'.
You have not accepted the license agreements of the following SDK components:
  [Android SDK Build-Tools 27.0.3].
  Before building your project, you need to accept the license agreements and complete the installation of the missing components using the Android Studio SDK Manager.
  Alternatively, to learn how to transfer the license agreements from one workstation to another, go to http://d.android.com/r/studio-ui/export-licenses.html
```

File: android/app/src/main/java/org/fossasia/openevent/common/events/NotificationDownloadEvent.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package org.fossasia.openevent.common.events;
+
+public class NotificationDownloadEvent extends DownloadEvent {
+
+    public NotificationDownloadEvent(boolean state) {
+        super(state);
+    }
+}
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/common/events/NotificationDownloadEvent.java]: This is unneeded I guess

File: android/app/src/main/java/org/fossasia/openevent/data/repository/RealmDataRepository.java [hunk 3]
```diff
@@ -574,6 +575,28 @@ public Completable saveEventDates(List<EventDates> eventDates) {
...
+        });
+    }
+
+    private void saveNotificationsInRealm(List<Notification> notifications) {
+        Realm realm = Realm.getDefaultInstance();
+        realm.executeTransaction(realm1 -> {
+            realm1.delete(Notification.class);
+            realm1.insertOrUpdate(notifications);
+        });
+        realm.close();
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/data/repository/RealmDataRepository.java]: Why a new instance

File: android/app/src/main/java/org/fossasia/openevent/core/notifications/NotificationsFragmentViewModel.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+                    .subscribeOn(Schedulers.io())
+                    .observeOn(AndroidSchedulers.mainThread())
+                    .subscribe( notifications -> {
+                        Timber.i("Downloaded notifications");
+                        realmRepo.saveNotifications(notifications);
+                        notificationDownloadResponse.setValue(true);
+                    }, throwable -> {
+                        Timber.e("Notification download failed");
+                        notificationDownloadResponse.setValue(false);
+                    }));
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/core/notifications/NotificationsFragmentViewModel.java]: This should be abstracted away from the VM, add a NotificationRepository

**Review by iamareebjamal — CHANGES_REQUESTED** [review]

File: android/app/src/main/java/org/fossasia/openevent/data/repository/RealmDataRepository.java (hunk 3 above)
Comment (dr0pdb) [reply android/app/src/main/java/org/fossasia/openevent/data/repository/RealmDataRepository.java]: Removed it.

**Comment by dr0pdb** [comment]
Changes: Created Notifications repository and deleted unused NotificationDownloadEvent class.

@iamareebjamal Please review.

File: android/app/src/main/java/org/fossasia/openevent/common/date/DateConverter.java [hunk 5]
```diff
@@ -169,4 +169,11 @@ public static String getRelativeTimeFromUTCTimeStamp(String timeStamp) throws Da
                 (timeCreatedDate.toInstant().toEpochMilli()),
                 System.currentTimeMillis(), android.text.format.DateUtils.SECOND_IN_MILLIS);
     }
+
+    public static String getRelativeTimeFromOffsetDateTime(String timeStamp) throws DateTimeParseException {
+        ZonedDateTime timeCreatedDate = ZonedDateTime.from(DateTimeFormatter.ISO_OFFSET_DATE_TIME.parse(timeStamp));
+        return (String) android.text.format.DateUtils.getRelativeTimeSpanString(
+                (timeCreatedDate.toInstant().toEpochMilli()),
+                System.currentTimeMillis(), android.text.format.DateUtils.SECOND_IN_MILLIS);
+    }
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/common/date/DateConverter.java]: We cannot unit test this. Any import of android.* should be forbidden for use in core classes. Create another class AndroidDateConverter, and add this method there, use any method present in DateConverter as a helper

File: android/app/src/main/java/org/fossasia/openevent/core/notifications/repository/NotificationsRepository.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+import org.json.JSONException;
+
+import io.reactivex.android.schedulers.AndroidSchedulers;
+import io.reactivex.disposables.Disposable;
+import io.reactivex.schedulers.Schedulers;
+import timber.log.Timber;
+
+public class NotificationsRepository {
+
+    public static Disposable downloadNotifications(MutableLiveData<Boolean> notificationDownloadResponse) {
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/core/notifications/repository/NotificationsRepository.java]: Don't make the method static, it creates problems during testing, and return an Observable and not a disposable. Responsibility of subscribing is on the caller

Comment (iamareebjamal) [reply android/app/src/main/java/org/fossasia/openevent/core/notifications/repository/NotificationsRepository.java]: Also. the Repository should have no idea about LiveData, only unidirectional flow of data should be present. Thus once you return Observable, you won't require LiveData

File: android/app/src/main/java/org/fossasia/openevent/data/repository/RealmDataRepository.java (hunk 3 above)
Comment (iamareebjamal) [reply android/app/src/main/java/org/fossasia/openevent/data/repository/RealmDataRepository.java]: realm.close() should be removed as well

**Review by iamareebjamal — CHANGES_REQUESTED** [review]

File: android/app/src/main/java/org/fossasia/openevent/common/date/DateConverter.java (hunk 5 above)
Comment (srv-twry) [reply android/app/src/main/java/org/fossasia/openevent/common/date/DateConverter.java]: The ```android.text.format.DateUtils.SECOND_IN_MILLIS``` is just ```1000L```. Can't i just add a constant in this DateConverter with value of ```1000L``` ?

**Comment by dr0pdb** [comment]
Changes: Created AndroidDateConverter class, NotificationsRepository returns Observable now.

@iamareebjamal Please review.

File: android/app/src/main/java/org/fossasia/openevent/core/notifications/NotificationsFragmentViewModel.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+    public LiveData<Boolean> downloadNotifications() {
+        if (notificationDownloadResponse == null) {
+            notificationDownloadResponse = new MutableLiveData<>();
+        }
+        Observable<List<Notification>> notificationsObservable = notificationsRepository.downloadNotifications();
+        if (notificationsObservable != null) {
+            compositeDisposable.add(notificationsObservable.subscribe(notifications -> {
+                Timber.i("Downloaded notifications");
+                notificationsRepository.saveNotifications(notifications);
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/core/notifications/NotificationsFragmentViewModel.java]: This should be responsibility of NotificationsRespository

**Review by iamareebjamal — CHANGES_REQUESTED** [review]

File: android/app/src/main/java/org/fossasia/openevent/core/notifications/NotificationsFragmentViewModel.java (hunk 7 above)
Comment (iamareebjamal) [reply android/app/src/main/java/org/fossasia/openevent/core/notifications/NotificationsFragmentViewModel.java]: Also add screenshot of what happens if there are no notifications

Comment (dr0pdb) [reply android/app/src/main/java/org/fossasia/openevent/core/notifications/NotificationsFragmentViewModel.java]: > This should be responsibility of NotificationsRespository

Sorry, i didn't get you. It is calling the repository only.

Comment (dr0pdb) [reply android/app/src/main/java/org/fossasia/openevent/core/notifications/NotificationsFragmentViewModel.java]: Did you mean chaining the call(using flatmap or something like that) ?

**Comment by dr0pdb** [comment]
In case of no notifications: 

screenshot_2018-02-09-04-39-15-71

File: android/app/src/main/java/org/fossasia/openevent/core/notifications/NotificationsFragmentViewModel.java (hunk 7 above)
Comment (iamareebjamal) [reply android/app/src/main/java/org/fossasia/openevent/core/notifications/NotificationsFragmentViewModel.java]: No, just use `onNext` in the repository to save the notifications in db

**Comment by dr0pdb** [comment]
Changes: Saving notifications in the repository itself.

@iamareebjamal Please review.

**Comment by iamareebjamal** [comment]
Please resolve conflicts @srv-twry

**Comment by dr0pdb** [comment]
@iamareebjamal Done.

**Review by iamareebjamal — APPROVED** [review]

**Pull request merged**